### PR TITLE
Converts Apc UI to Tgui [TGUI]

### PR DESCRIPTION
**Pull request by 56kyle**
## About the PR 
Converts the Apc UI largely found in /obj/machinery/power/apc/proc/interacted and /obj/machinery/power/apc/proc/Topic into tgui. Does so by following along the events in the Topic proc and then turning them into ui_act based events that have callback procs defined in apc.dm. Generally speaking tries to maintain the same logic as much as possible, but implements some portions into the javascript and also adds some helper functions to reduce existing repeated code.

Additionally removes the Topic prompt entirely and nullifies the majority of the interacted proc. As for the actual inferno components, unifies the Apc's main panel and the access panel into one ui window. Generally speaking tries to maintain as close to the original UI's look as possible to make it easier to verify that all is the same. (I plan on eventually making it prettier, but for now functionality staying the same is the number one priority).
Edit: Made it prettier

In terms of events, the only large change would be to how wires are handled. The mend and cut events were separated, as well as the wires cut state was split up for each wire.

Image of UI

### Things to look out for
Overall I tested most possible cases, however I was unable to figure out how to test playing as the AI in a private server. Any help testing this would be greatly appreciated.

Also there is definitely still some fat that can be trimmed from the old ui

## Why's this needed? 
Wires open up in a different window
Neither window updates in response to changes
I use Telesci a lot and these factors test my sanity

## Changelog
```changelog
(u)Taylor Kellogg
(+)Converted the APC's UI to TGUI
```

### CODE REVIEW

**Comment by BatElite** [comment]
Got a screenshto of the UI in action?

**Comment by 56kyle** [comment]
ye sorry for delay, will grab screenshots rn

File: tgui/packages/tgui/interfaces/Apc.js [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+    act("onChargeModeChange", { chargemode });
+  };
+
+  const onCoverLockedChange = (coverlocked) => {
+    act("onCoverLockedChange", { coverlocked });
+  };
+
+  const onOverload = () => {
+    act("onOverload", {});
+  };
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/Apc.js:265]: inline these

**Comment by Azrun** [comment]
Would prefer to see power channel buttons vertically aligned. And a better indication of the current state?

**Comment by BatElite** [comment]
per-channel load is also missing atm

**Comment by ZeWaka** [comment]
swipe id card should also probably be centered and bolded or something

**Comment by 56kyle** [comment]
> And a better indication of the current state?

ye definitely agree, was trying to get them to do so for a while but is a bit akward due to how I structured the components. Will revisit and do

**Comment by 56kyle** [comment]
> per-channel load is also missing atm

ah ty, will fix

**Comment by 56kyle** [comment]
> swipe id card should also probably be centered and bolded or something

alrighty, will do

**Comment by frawhst** [comment]
!merge_upstream

**Comment by 56kyle** [comment]
Alrighty there we go, this should be good to go now in terms of aforementioned changes

File: code/modules/power/apc.dm [hunk 2]
```diff
@@ -583,150 +583,319 @@ var/zapLimiter = 0
...
+		"net_id" = net_id,
+		"area_name" = area ? area.name : "Unknown",
+		"area_requires_power" = area ? area.requires_power : null,
+	)
+
+/obj/machinery/power/apc/ui_data(mob/user)
+	. = list(
+		"cell_type" = cell_type, // 0=no cell, 1=regular, 2=high-cap (x5) <- old, now it's just 0=no cell, otherwise dictate cellcapacity by changing this value. 1 used to be 1000, 2 was 2500
+		"cell_percent" = cell ? cell.percent() : null,
+		"cell_present" = cell ? 1 : 0,
```
Comment (ZeWaka) [inline code/modules/power/apc.dm:604]:
```suggestion
		"cell_present" = !!cell,
```

File: code/modules/power/apc.dm [hunk 3]
```diff
@@ -583,150 +583,319 @@ var/zapLimiter = 0
...
-			t1 += "<br>"
-		t1 += text("<br><br>[(src.locked ? "The APC is locked." : "The APC is unlocked.")]<br><br>[(src.shorted ? "The APCs power has been shorted." : "The APC is working properly!")]<br><br>[(src.aidisabled ? "The 'AI control allowed' light is off." : "The 'AI control allowed' light is on.")]")
-		t1 += text("<p><a href='?src=\ref[src];close2=1'>Close</a></p><br>")
-		user.Browse(t1, "window=apcwires")
-		onclose(user, "apcwires")
-
-	if (can_access_remotely(user) && src.aidisabled)
-		boutput(user, "AI control for this APC interface has been disabled.")
-		user.Browse(null, "window=apc")
+	if (usr.getStatusDuration("stunned") || usr.getStatusDuration("weakened") || usr.stat)
```
Comment (ZeWaka) [inline code/modules/power/apc.dm:648]: can_act

File: code/modules/power/apc.dm [hunk 4]
```diff
@@ -583,150 +583,319 @@ var/zapLimiter = 0
...
 		return
 
-	src.add_dialog(user)
-	var/t = "<TT><B>Area Power Controller</B> ([area.name])<HR>"
+	if ((in_interact_range(src, usr) && istype(src.loc, /turf))||(issilicon(usr) || isAI(usr)))
```
Comment (ZeWaka) [inline code/modules/power/apc.dm:651]: spacing

File: tgui/packages/tgui/interfaces/Apc/index.js [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        </LabeledList.Item>
+      );
+    }
+  };
+
+  const renderPoweredAreaApc = () => {
+    return (
+      <Window title="Area Power Controller" width={400} height={data["wiresexposed"] ? 680 : 420}>
+        <Window.Content>
+          <Section title={"Area Power Controller (" + area_name + ")"}>
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/Apc/index.js:205]: do we really need to say area power controller twice

File: tgui/packages/tgui/interfaces/Apc/index.js [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        </Window.Content>
+      </Window>
+    );
+  };
+
+  const renderUnPoweredAreaApc = () => {
+    return (
+      <Window title="Area Power Controller" width={400} height={wiresexposed ? 500 : 350}>
+        <Window.Content>
+          <Section title={"Area Power Controller (" + area_name + ")"}>
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/Apc/index.js:250]: x

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/modules/power/apc.dm (hunk 3 above)
Comment (56kyle) [reply code/modules/power/apc.dm]: ah nice, ty didn't know that existed

File: code/modules/power/apc.dm (hunk 4 above)
Comment (56kyle) [reply code/modules/power/apc.dm]: fixed

File: tgui/packages/tgui/interfaces/Apc/index.js (hunk 5 above)
Comment (56kyle) [reply tgui/packages/tgui/interfaces/Apc/index.js]: replaced with just the name of area in second instance

File: tgui/packages/tgui/interfaces/Apc/index.js (hunk 6 above)
Comment (56kyle) [reply tgui/packages/tgui/interfaces/Apc/index.js:222]: assuming this is just referencing location of second occurrance?

**Comment by 56kyle** [comment]
Just an update, all changes requested by ZeWaka are done, just waiting on re-review

**Comment by stonepillars** [comment]
Why not use the On/Off components used in the SMES TGUI for the Main breaker and charging?

**Pull request merged**